precompiles/dispatch-lockdrop/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+
+extern crate alloc;
+
+use alloc::format;
+use core::marker::PhantomData;
+use fp_evm::{ExitError, PrecompileFailure, PrecompileHandle};
+use frame_support::pallet_prelude::IsType;
+use frame_support::{codec::DecodeLimit as _, dispatch::Pays, traits::Get};
+use frame_support::{
+    dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+    traits::ConstU32,
+};
+use frame_system::Config;
+use pallet_evm::GasWeightMapping;
+use pallet_evm_precompile_dispatch::DispatchValidateT;
+use parity_scale_codec::Encode;
+use precompile_utils::prelude::{BoundedBytes, UnboundedBytes};
+use precompile_utils::{keccak256, EvmResult};
+use sp_core::ecdsa::Signature;
+use sp_core::{crypto::AccountId32, H160, H256};
+use sp_core::{ecdsa, U256};
+use sp_io::hashing::keccak_256;
+use sp_runtime::traits::Zero;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub const LOG_TARGET: &str = "dispatch-lockdrop-precompile";
+
+// ECDSA signature bytes
+type ECDSASignatureBytes = ConstU32<65>;
+
+// `DecodeLimit` specifies the max depth a call can use when decoding, as unbounded depth
+// can be used to overflow the stack.
+// Default value is 8, which is the same as in XCM call decoding.
+pub struct DispatchLockdrop<Runtime, DispatchValidator, DecodeLimit = ConstU32<8>>(
+    PhantomData<(Runtime, DispatchValidator, DecodeLimit)>,
+);
+
+#[precompile_utils::precompile]
+impl<Runtime, DispatchValidator, DecodeLimit>
+    DispatchLockdrop<Runtime, DispatchValidator, DecodeLimit>
+where
+    Runtime: pallet_evm::Config,
+    <Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin: From<Option<Runtime::AccountId>>,
+    Runtime::RuntimeCall: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+    <Runtime as Config>::AccountId: IsType<AccountId32>,
+    <Runtime as Config>::AccountId: From<[u8; 32]>,
+    DispatchValidator:
+        DispatchValidateT<<Runtime as Config>::AccountId, <Runtime as Config>::RuntimeCall>,
+    DecodeLimit: Get<u32>,
+{
+    #[precompile::public("dispatch_lockdrop_call(bytes,bytes32,bytes)")]
+    fn dispatch_lockdrop_call(
+        handle: &mut impl PrecompileHandle,
+        call: UnboundedBytes,
+        account_id: H256,
+        signature: BoundedBytes<ECDSASignatureBytes>,
+    ) -> EvmResult<bool> {
+        log::trace!(
+            target: LOG_TARGET,
+            "raw arguments: call: {:?}, account_id: {:?}, signature: {:?}",
+            call,
+            account_id,
+            signature
+        );
+
+        let caller: H160 = handle.context().caller.into();
+        let input: Vec<u8> = call.into();
+        let signature_bytes: Vec<u8> = signature.into();
+        let account_id = AccountId32::new(account_id.into()).into();
+
+        // 1. Decode the call
+        let call = unwrap_or_err!(
+            Runtime::RuntimeCall::decode_with_depth_limit(DecodeLimit::get(), &mut &*input).ok(),
+            "Error: could not decode call"
+        );
```

Comment:
There is `revert()` helper method in utils for this, more idiomatic. (for all the occurrences)
```rust
let call =  Runtime::RuntimeCall::decode_with_depth_limit(DecodeLimit::get(), &mut &*input).map_err(|_| revert("could not decode call"))?;
```

Reply:
Great suggestion! Thanks